orca-core/src/main/java/com/netflix/spinnaker/orca/notifications/scheduling/TopApplicationExecutionCleanupPollingNotificationAgent.java

```diff
@@ -139,28 +143,38 @@ private void cleanup(Observable<PipelineExecution> observable, String applicatio
         observable.filter(filter).map(mapper).toList().toBlocking().single();
     executions.sort(comparing(a -> (Long) Optional.ofNullable(a.get("startTime")).orElse(0L)));
     if (executions.size() > threshold) {
-      executions
-          .subList(0, (executions.size() - threshold))
-          .forEach(
-              it -> {
-                Long startTime =
-                    Optional.ofNullable((Long) it.get("startTime"))
-                        .orElseGet(() -> (Long) it.get("buildTime"));
-                log.info(
-                    "Deleting {} execution {} (startTime: {}, application: {}, pipelineConfigId: {}, status: {})",
-                    type,
-                    it.get("id"),
-                    startTime != null ? Instant.ofEpochMilli(startTime) : null,
-                    application,
-                    it.get("pipelineConfigId"),
-                    it.get("status"));
-                if (type.equals("orchestration")) {
-                  executionRepository.delete(ORCHESTRATION, (String) it.get("id"));
-                  registry.counter(deleteCountId.withTag("application", application)).increment();
-                } else {
-                  throw new IllegalArgumentException(format("Unsupported type '%s'", type));
-                }
-              });
+      List<? extends Map> removingPipelineExecutions =
+          executions.subList(0, (executions.size() - threshold));
+
+      List<String> removingPipelineExecutionIds =
+          removingPipelineExecutions.stream()
+              .map(pipelineExecution -> (String) pipelineExecution.get("id"))
+              .collect(Collectors.toList());
+
+      pipelineDependencyCleanupOperators.forEach(
+          pipelineDependencyCleanupOperator ->
+              pipelineDependencyCleanupOperator.cleanup(removingPipelineExecutionIds));
+
+      removingPipelineExecutions.forEach(
+          it -> {
+            Long startTime =
+                Optional.ofNullable((Long) it.get("startTime"))
+                    .orElseGet(() -> (Long) it.get("buildTime"));
+            log.info(
+                "Deleting {} execution {} (startTime: {}, application: {}, pipelineConfigId: {}, status: {})",
+                type,
+                it.get("id"),
+                startTime != null ? Instant.ofEpochMilli(startTime) : null,
+                application,
+                it.get("pipelineConfigId"),
+                it.get("status"));
+            if (type.equals("orchestration")) {
```

Comment:
Not sure if `type` can be null at this point, but just to avoid a possible NPE maybe you can switch the equality check to `"orchestration".equals(type).`